### DISCUSSION TITLE
Significant Increase in INTERNAL_EDGE_FUNCTION_UNHANDLED_ERROR Errors

### DISCUSSION THREAD

**Discussion by k9ordon** [vercel/community#3764] [Bug Report]
### Summary

Over the recent period, we've experienced a significant uptick in the `INTERNAL_EDGE_FUNCTION_UNHANDLED_ERROR` errors.

### Middleware Setup:
Our application uses a straightforward Vercel middleware without incorporating any external dependencies. The middleware's main functionality is to compare process.env.APIKEY with the APIKEY header from incoming requests. In case of a mismatch, a 401 status is promptly returned. Despite occasional errors in the past, their frequency remained relatively low. However, from August 24, 2023, 12:00 CEST onward, we've noticed a stark increase in the error rate associated with this middleware.

### Error Log:

```
[POST] [middleware: "src/middleware"] /api reason=INTERNAL_EDGE_FUNCTION_UNHANDLED_ERROR, status=500, upstream_status=503, user_error=false
```

### Middleware Code:

```ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

export function middleware(request: NextRequest) {
  if (process.env.NODE_ENV === 'production') {
    const envAPIKey = process.env.APIKEY;
    const apiKey = request.headers.get('apikey')?.toString();

    if (!envAPIKey || !apiKey || envAPIKey !== apiKey) {
      return new NextResponse('Invalid API-Key provided.', {
        status: 401,
      });
    }
  }
}

export const config = {
  matcher: '/api',
};
```

### Error Frequency:
A graph showcasing the error count of INTERNAL_EDGE_FUNCTION_UNHANDLED_ERROR for the past 30 days
![Bildschirmfoto 2023-08-25 um 12 38 31](https://github.com/vercel/community/assets/102591/b47def96-b777-4fad-9d3f-3ec0766cb2ed)
It's important to note that we haven't made any alterations to the middleware's code during this time. Additionally, the depicted errors represent only a small fraction of the requests that are experiencing issues.

Despite reaching out to support on July 5, we've yet to receive a response. Given the escalating nature of this issue, we're eager to resolve it as soon as possible.

Is anyone else familiar with this issue? Has anyone found solutions or have ideas regarding this? Any feedback or shared experience would be greatly appreciated.

---

### Steps to Reproduce

1. Implement the provided middleware code in a Vercel application.
2. Ensure that the setup correctly compares the process.env.APIKEY with the incoming APIKEY header for validation.
3. Simulate or initiate a very high volume of requests to the application. You may require tools or scripts to send a burst of requests concurrently.
4. Monitor the application for any instances of the INTERNAL_EDGE_FUNCTION_UNHANDLED_ERROR error.
5. Take note of any occurrences in different edge regions, as we've observed this issue happening across almost all of them.

It's worth noting that reproducing this issue has proven challenging. However, given the high frequency of requests and the middleware configuration, the error should manifest eventually.

**Comment by mmachatschek** [comment]
Deployment ID for support 2wu8NwaVAH1sFgvfrfcmn4pqr1pZ

**Comment by amyegan** [comment]
Thanks for reporting this. I shared it with the Edge team and we'll let you know when we have more info 🙇

**Comment by amyegan** [comment]
It looks like there was a network issue causing the errors which resolved over the weekend. 

Please let me know if see this problem again 🙏

**Comment by k9ordon** [comment]
Yes, we are still seeing the errors persist. While they are not spread throughout the day as they were over the weekend, we've noticed two spikes in the last two days. I've attached the error graph for `INTERNAL_EDGE_FUNCTION_UNHANDLED_ERROR` from the past 7 days for reference. The support team has also reached out to us in the meantime.

![Bildschirmfoto 2023-08-30 um 11 35 47](https://github.com/vercel/community/assets/102591/349093d0-eea7-492c-b129-c6cbb7a25c59)

**Comment by k9ordon** [comment]
@amyegan Are there again any network issues? We now see the same thing happening with 504 EDGE_FUNCTION_INVOCATION_TIMEOUT errors 🤔. It's reminiscent of the network challenges we faced 2 weeks ago.

24h Graph: ![Bildschirmfoto 2023-09-08 um 09 54 37](https://github.com/vercel/community/assets/102591/b81f14c8-d728-4505-80ae-a829afd27b59)

**Reply by amyegan** [reply]
Thanks for sharing this. I asked the Edge team for help investigating 🙏

**Comment by k9ordon** [comment]
The same issue surfaced again in production yesterday. We have a very simple edge middleware that performs a straightforward comparison between process environment variables and request header variables. This specific middleware, as well as many of our other edge functions, experience this issue on a daily basis, albeit in small numbers. However, the frequency of occurrences spiked yesterday. Attached are the 24-hour and 30-day graphs for reference.

![24h](https://github.com/vercel/community/assets/102591/3b8fa561-f847-4f3c-9092-36a5a3d0501f)
![30d](https://github.com/vercel/community/assets/102591/c3750360-8fcd-4491-9825-a223ba1f23e1)

**Reply by amyegan** [reply]
I don't see any similar reports from the past few days but I'll keep digging. Was the problem confined to that day, or is it still happening?

**Reply by mmachatschek** [reply]
@amyegan the errors are still there (see the report of the last 2 days, below), the issues already have been acknowledged by the support team in case `00153758`

![image](https://github.com/vercel/community/assets/10237069/9b91a14d-d602-4091-9c16-611eef226ca2)
